File: Week27 Codes/5.Post Quantum Cryptography/signing.cpp
#include <botan/auto_rng.h>
#include <botan/ec_group.h>
#include <botan/ecdsa.h>
#include <botan/hex.h>
#include <botan/pubkey.h>
#include <iostream>
#include <botan/dilithium.h>

int main() {
Botan::AutoSeeded_RNG rng;
// Generate ECDSA keypair
Botan::Dilithium_PrivateKey key(rng, Botan::DilithiumMode::Mode::Dilithium4x4);
const std::string message("This is a tasty burger!");
// sign data
Botan::PK_Signer signer(key,rng,"");
signer.update(message);
std::vector<uint8_t> signature = signer.signature(rng);
std::cout << "Signature:\n" << Botan::hex_encode(signature);
// now verify the signature
Botan::PK_Verifier verifier(key, "");
verifier.update(message);
std::cout << "is " << (verifier.check_signature(signature) ? "valid" : "invalid")<<std::endl;
return 0;

}

File: Week27 Codes/5.Post Quantum Cryptography/signing.py
import botan3 as botan
from base64 import b64encode

def main():
    rng = botan.RandomNumberGenerator()

    # Generate RSA private key
    dilithium_priv_key = botan.PrivateKey.create("Dilithium", "", rng)

    # Extract public key from private key
    dilithium_pub_key = dilithium_priv_key.get_public_key()

    # Message to sign
    message = "Hello, this is a signed message."

    # Sign the message
    signer = botan.PKSign(dilithium_priv_key, '')
    signer.update(message)
    signature = signer.finish(rng)

    # Hex encode the signature
    hex_signature = b64encode(signature).decode()

    # Verify the signature
    verifier = botan.PKVerify(dilithium_pub_key, '')
    verifier.update(message)
    valid = verifier.check_signature(signature)

    # Output the results
    print("Message:", message)
    print("Signature:", hex_signature)
    print("Verification:", "valid" if valid else "invalid")

if __name__ == "__main__":
    main()